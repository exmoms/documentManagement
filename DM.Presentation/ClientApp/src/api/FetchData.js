/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
