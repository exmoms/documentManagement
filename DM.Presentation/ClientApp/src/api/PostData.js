/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
