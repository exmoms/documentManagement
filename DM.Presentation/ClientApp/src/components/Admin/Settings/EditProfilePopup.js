/**
 * @file Contains BrowseAggregatedDocsPopUp component implementaion
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @version 1.0
 */

import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import Paper from "@material-ui/core/Paper";
import { withStyles } from "@material-ui/core/styles";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableContainer from "@material-ui/core/TableContainer";
import TableRow from "@material-ui/core/TableRow";
import Typography from "@material-ui/core/Typography";
import React, { Component } from "react";
import { withTranslation } from "react-i18next";
import { fetchData } from "../../../api/FetchData";
import { postDataToAPI } from "../../../api/PostData";
import { Notification } from "../Notifications";
import EditUserProfileByAdmin from "./EditUserProfileByAdmin";

const styles = (theme) => ({
  paper: {
    maxWidth: "auto",
  },
  margin: {
    margin: theme.spacing(2),
  },
});

class EditProfilePopup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user: this.props.user,
      userProfile: null,
      errors: {
        confirmErrors: "",
        newPasswordErrors: "",
        submissionErrors: "",
      },
      showSuccess: false,
    };
    this.updateParentState = this.props.action;
  }
  t = this.props["t"];

  componentDidUpdate(prevProps) {
    if (prevProps.user.userId !== this.props.user.userId) {
      fetchData(`/api/user/${this.props.user.userId}`).then((profileData) => {
        this.setState({ userProfile: profileData });
      });
    }
  }

  toggleShow = (show) => {
    this.updateParentState(show);
  };

  handelSubmit = (event) => {
    event.preventDefault();
    let email = document.getElementById("edit-profile-by-admin-email").value;
    let tel = document.getElementById("edit-profile-by-admin-tel").value;
    let newPassword = document.getElementById(
      "edit-profile-by-admin-outlined-adornment-password"
    ).value;
    let confirmPassword = document.getElementById(
      "edit-profile-by-admin-confirmPassword"
    ).value;
    let validPassword = false;
    if (newPassword === "") {
      validPassword = newPassword === confirmPassword;
    } else {
      validPassword =
        newPassword === confirmPassword &&
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^\w]).{8,}$/i.test(newPassword);
    }

    if (validPassword) {
      const formData = {
        userId: this.props.user.userId,
        email: email,
        phoneNumber: tel,
        NewPassword: newPassword,
        ConfirmPassword: confirmPassword,
      };
      postDataToAPI("/api/user/EditUserProfileByAdmin", formData)
        .then(
          (res) => {
            if (res.ok) {
              this.setState({
                errors: {
                  confirmErrors: "",
                  newPasswordErrors: "",
                  submissionErrors: "",
                },
                showSuccess: true,
              });
            }
            return res.json();
          },
          (error) => console.log(error)
        )
        .then(
          (data) => {
            this.setState({
              errors: {
                confirmErrors: "",
                newPasswordErrors: "",
                submissionErrors: data.error,
              },
              showSuccess: true,
            });
          },
          (error) => console.log(error)
        );

      this.toggleShow(false);
    } else {
      this.setState({
        errors: {
          confirmErrors:
            newPassword === confirmPassword
              ? ""
              : this.t("password_mismatchingWithComfirm"),
          newPasswordErrors: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^\w]).{8,}$/i.test(
            newPassword
          )
            ? ""
            : this.t("password_formateError"),
        },
      });
    }
  };

  render() {
    const { classes } = this.props;
    return (
      <div>
        <Dialog
          position="centerCenter"
          open={this.props.show}
          onClose={() => this.toggleShow(false)}
          aria-labelledby="form-dialog-title"
          className={classes.paper}
        >
          <DialogTitle id="form-dialog-title">
            <Typography variant="h6" gutterBottom align="left">
              {this.t("edit_user_profile")} : {this.props.user.userName}
            </Typography>
          </DialogTitle>
          <form
            className={classes.form}
            id="edit-profile-form"
            onSubmit={this.handelSubmit}
          >
            <DialogContent>
              <TableContainer component={Paper}>
                <Table className={classes.table} aria-label="simple table">
                  <TableBody>
                    <TableRow key="search-form">
                      <TableCell>
                        <EditUserProfileByAdmin
                          profileData={this.state.userProfile}
                          errors={this.state.errors}
                        />
                      </TableCell>
                    </TableRow>
                  </TableBody>
                </Table>
              </TableContainer>
            </DialogContent>
            <DialogActions>
              <Button
                id="submit-btn"
                type="submit"
                // variant="contained"
                color="primary"
                className={classes.submit}
              >
                {this.t("submit_changes")}
              </Button>
              <Button
                id="CancelButton"
                onClick={() => this.toggleShow(false)}
                color="primary"
              >
                {this.t("button_cancel")}
              </Button>
            </DialogActions>
          </form>
        </Dialog>
        <Notification
          id="notification"
          open={this.state.showSuccess}
          setOpen={(show) => this.setState({ showSuccess: show })}
          error={!!this.state.errors.submissionErrors}
          errorMessage={this.state.errors.submissionErrors}
        />
      </div>
    );
  }
}

export default withStyles(styles)(withTranslation()(EditProfilePopup));
