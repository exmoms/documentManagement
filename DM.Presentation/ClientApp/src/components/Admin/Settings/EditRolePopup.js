/**
 * @file Contains EditRolePopup component implementaion
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @version 1.0
 */

import {
  Button,
  FormControl,
  Input,
  InputLabel,
  MenuItem,
  Select,
} from "@material-ui/core";

import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import { withStyles } from "@material-ui/core/styles";
import React, { Component } from "react";
import { withTranslation } from "react-i18next";
import { postDataToAPI } from "../../../api/PostData";
import { Notification } from "../Notifications";

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 400,
    },
  },
};

const styles = (theme) => ({
  paper: {
    maxWidth: "auto",
    marginTop: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
  },
  margin: {
    margin: theme.spacing(2),
  },
});

class EditRolePopup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userId: this.props.user.userId,
      selectedRole: "",
      currentRole: "",
      availableRoles: ["Admin", "User"],
      error: false,
      errorMessage: [],
      showSuccess: false,
    };
    this.updateParentState = this.props.action;
  }
  t = this.props["t"];

  componentDidUpdate(prevProps) {
    if (
      prevProps.user.userId !== this.props.user.userId ||
      JSON.stringify(prevProps.user.roles) !==
        JSON.stringify(this.props.user.roles)
    ) {
      this.setState({
        userId: this.props.user.userId,
        selectedRole: this.props.user.roles[0],
        currentRole: this.props.user.roles[0],
      });
    }
  }

  toggleShow = (show, flag) => {
    this.updateParentState(show, flag);
  };

  handleChange = (event) => {
    this.setState({ selectedRole: event.target.value });
  };

  handelSubmit = async (event) => {
    event.preventDefault();

    const formData = {
      userId: this.props.user.userId,
      AssignedRoles: [],
      UnassignedRoles: [],
    };

    if (this.state.currentRole !== this.state.selectedRole) {
      formData.AssignedRoles.push(this.state.selectedRole);
      formData.UnassignedRoles.push(this.state.currentRole);
    }

    let response = await postDataToAPI("/api/user/EditUserRoles", formData);
    if (response.ok) {
      this.setState({ showSuccess: true });
    } else {
      let json = await response.json();
      this.setState({
        showSuccess: true,
        error: true,
        errorMessage: json.errorMessage,
      });
    }
    this.toggleShow(false, response.ok);
  };

  render() {
    const { classes } = this.props;
    let lang = localStorage.getItem("i18nextLng");
    return (
      <div>
        <Dialog
          position="centerCenter"
          open={this.props.show}
          onClose={() => this.toggleShow(false, false)}
          aria-labelledby="form-dialog-title"
          className={classes.paper}
        >
          <div
            style={
              lang === "ar" ? { textAlign: "right" } : { textAlign: "left" }
            }
          >
            <DialogTitle id="form-dialog-title">
              {this.t("edit_user_role")} : {this.props.user.userName}
            </DialogTitle>
          </div>

          <DialogContent>
            <div className={classes.paper}>
              <FormControl style={{ minWidth: 400, minHeight: 120 }}>
                <InputLabel id="demo-mutiple-chip-label">
                  {this.t("role_label")}
                </InputLabel>
                <Select
                  labelId="demo-mutiple-chip-label"
                  id="demo-mutiple-chip"
                  autoWidth
                  value={this.state.selectedRole}
                  onChange={this.handleChange}
                  input={<Input id="select-multiple-chip" />}
                  MenuProps={MenuProps}
                >
                  {this.state.availableRoles.map((name, index) => (
                    <MenuItem key={index} value={name}>
                      {name}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </div>
          </DialogContent>
          <DialogActions>
            <Button
              id="submit-btn"
              onClick={this.handelSubmit}
              color="primary"
              className={classes.submit}
            >
              {this.t("submit_changes")}
            </Button>
            <Button
              id="CancelButton"
              onClick={() => this.toggleShow(false, false)}
              color="primary"
            >
              {this.t("button_cancel")}
            </Button>
          </DialogActions>
        </Dialog>
        <Notification
          id="notification"
          open={this.state.showSuccess}
          setOpen={(show) => this.setState({ showSuccess: show })}
          error={!!this.state.error}
          errorMessage={this.state.errorMessage}
        />
      </div>
    );
  }
}

export default withStyles(styles)(withTranslation()(EditRolePopup));
