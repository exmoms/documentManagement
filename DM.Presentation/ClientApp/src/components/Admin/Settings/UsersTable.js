import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import { default as Check } from "@material-ui/icons/Check";
import ChevronLeft from "@material-ui/icons/ChevronLeft";
import ChevronRight from "@material-ui/icons/ChevronRight";
import Clear from "@material-ui/icons/Clear";
import DeleteOutline from "@material-ui/icons/DeleteOutline";
import { default as Edit, default as EditIcon } from "@material-ui/icons/Edit";
import FilterList from "@material-ui/icons/FilterList";
import FirstPage from "@material-ui/icons/FirstPage";
import LastPage from "@material-ui/icons/LastPage";
import Remove from "@material-ui/icons/Remove";
import SaveAlt from "@material-ui/icons/SaveAlt";
import Search from "@material-ui/icons/Search";
import ViewColumn from "@material-ui/icons/ViewColumn";
import UpdateIcon from "@material-ui/icons/Update";
import MaterialTable from "material-table";
import React, { Component, forwardRef } from "react";
import { withTranslation } from "react-i18next";
import { fetchData } from "../../../api/FetchData";
import TableBodyRow from "../../Shared/materialTableRow";
import TableBodyToolbar from "../../Shared/materialTableToolbar";
import EditProfilePopup from "./EditProfilePopup";
import EditRolePopup from "./EditRolePopup";

const tableIcons = {
  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
  Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
  DetailPanel: forwardRef((props, ref) => (
    <ChevronRight {...props} ref={ref} />
  )),
  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  PreviousPage: forwardRef((props, ref) => (
    <ChevronLeft {...props} ref={ref} />
  )),
  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
  SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
  ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
  ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />),
};

class UsersTable extends Component {
  getAllLessUsersURL = "/api/user/lessRoleUsers";
  initUser = {
    userId: 0,
    userName: "",
  };

  constructor(props) {
    super(props);
    this.state = {
      option: 1,
      data: [],
      user: this.initUser,
      show: false,
      showChangeRolePopup: false,
    };
  }

  componentDidMount() {
    fetchData(this.getAllLessUsersURL).then(
      (data) => {
        this.setState({
          data: data,
        });
      },
      (error) => console.log(error)
    );
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.userTableNeedUpdate &&
      this.props.userTableNeedUpdate !== prevProps.userTableNeedUpdate
    ) {
      fetchData(this.getAllLessUsersURL).then(
        (data) => {
          this.setState({
            data: data,
          });
        },
        (error) => console.log(error)
      );
    }
  }

  // translation object
  t = this.props["t"];

  editUserProfile = {
    icon: EditIcon,
    tooltip: this.t("edit_user_profile"),
    position: "row",
    onClick: (event, rowData) => {
      this.setState({ user: rowData, show: true });
    },
  };

  changeRole = {
    icon: UpdateIcon,
    tooltip: this.t("edit_user_role"),
    position: "row",
    onClick: (event, rowData) => {
      this.setState({ user: rowData, showChangeRolePopup: true });
    },
  };

  composeTableColumns = () => {
    let columns = [];

    columns = [
      {
        title: this.t("userid"),
        field: "userId",
      },
      { title: this.t("username"), field: "userName" },
      {
        title: this.t("role_label"),
        field: "role",
        hidden: true,
      },
    ];
    return columns;
  };

  composeTableActions = () => {
    let actions = [];
    let userRole = localStorage.getItem("userRole");
    if (userRole !== null && userRole !== undefined) {
      actions =
        userRole === "SuperAdmin"
          ? [this.editUserProfile, this.changeRole]
          : [this.editUserProfile];
    } else {
      actions = [this.editUserProfile];
    }

    return actions;
  };

  showPopup = (show) => {
    this.setState({ show: show });
  };

  showChangeRolePopup = (show, flag) => {
    if (flag) {
      fetchData(this.getAllLessUsersURL).then(
        (data) => {
          this.setState({
            data: data,
            showChangeRolePopup: show,
          });
        },
        (error) => console.log(error)
      );
    } else {
      this.setState({ showChangeRolePopup: show });
    }
  };

  render() {
    const { user, show, showChangeRolePopup } = this.state;
    return (
      <div style={{ maxWidth: "100%" }}>
        <MaterialTable
          title={this.t("browse_users")}
          icons={tableIcons}
          columns={this.composeTableColumns()}
          components={{
            Row: TableBodyRow,
            Toolbar: TableBodyToolbar,
          }}
          data={this.state.data}
          actions={this.composeTableActions()}
          options={{
            actionsColumnIndex: -1,
            exportButton: false,
          }}
          localization={{
            pagination: {
              labelRowsSelect: this.t("rows"),
              firstAriaLabel: this.t("first_page"),
              firstTooltip: this.t("first_page"),
              previousAriaLabel: this.t("previous_page"),
              previousTooltip: this.t("previous_page"),
              nextAriaLabel: this.t("next_page"),
              nextTooltip: this.t("next_page"),
              lastAriaLabel: this.t("last_page"),
              lastTooltip: this.t("last_page"),
            },
            toolbar: {
              nRowsSelected: "{0}" + this.t("selected_row"),
              searchTooltip: this.t("search"),
              searchPlaceholder: this.t("search"),
              exportTitle: this.t("export_as_csv"),
            },
            header: {
              actions: this.t("actions"),
            },
            body: {
              emptyDataSourceMessage: this.t("no_records"),
            },
          }}
        />
        <EditProfilePopup show={show} action={this.showPopup} user={user} />
        <EditRolePopup
          show={showChangeRolePopup}
          action={this.showChangeRolePopup}
          user={user}
        />
      </div>
    );
  }
}

export default withTranslation()(UsersTable);
