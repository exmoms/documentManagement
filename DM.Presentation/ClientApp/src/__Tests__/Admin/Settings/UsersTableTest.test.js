import React from "react";
import ReactDom from "react-dom";
import UsersTable from "../../../components/Admin/Settings/UsersTable";
import EditProfilePopup from "../../../components/Admin/Settings/EditProfilePopup";
import EditRolePopup from "../../../components/Admin/Settings/EditRolePopup";
import { cleanup } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import TextField from "@material-ui/core/TextField";
import { act } from "react-dom/test-utils";
import Button from "@material-ui/core/Button";
import { BrowserRouter as Router } from "react-router-dom";
import * as FetchMock from "../../../api/FetchData";
import * as PostDataMock from "../../../api/PostData";
import Typography from "@material-ui/core/Typography";
import Dialog from "@material-ui/core/Dialog";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import LinearProgress from "@material-ui/core/LinearProgress";
import MaterialTable from "material-table";
afterEach(cleanup);
describe("UsersTable", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(
        <UsersTable />,
      div
    );
  });

  test("matches snapshot", () => {
    const wrapper = shallow(
        <UsersTable/>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("intialization UsersTable", () => {
    let  wrapper = mount(<UsersTable/>) ;
    expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual([]);
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].title).toBe("userid");
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].field).toBe("userId");

    expect(wrapper.find(MaterialTable).at(0).props().columns[1].title).toBe("username");
    expect(wrapper.find(MaterialTable).at(0).props().columns[1].field).toBe("userName");

    expect(wrapper.find(MaterialTable).at(0).props().columns[2].title).toBe("role_label");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].field).toBe("role");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].hidden).toBeTruthy();

    expect(wrapper.find(EditProfilePopup).at(0).props().show).toBeFalsy();
    expect(wrapper.find(EditProfilePopup).at(0).props().user).toStrictEqual({
        userId: 0,
        userName: "",
      });
  });
   
  test("componentDidMount", () => {
    const mockSuccessResponse = [{id:1 , name:"Name A"} , {id:2 , name:"Name B"}];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise);
    let wrapper;
    return mockJsonPromise.then(() => {
      wrapper = mount(<UsersTable />) ;
    }).then(() => {
        wrapper.render();
        wrapper.update();
        expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock.mock.calls[0][0]).toBe("/api/user/lessRoleUsers");
      expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual(mockSuccessResponse);
    expect(wrapper.find(EditProfilePopup).at(0).props().show).toBeFalsy();
    expect(wrapper.find(EditProfilePopup).at(0).props().user).toStrictEqual({
        userId: 0,
        userName: "",
      });
    });
});

test("showPopup functionality", () => {
    let  wrapper = mount(<UsersTable/>) ;
    expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual([]);
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].title).toBe("userid");
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].field).toBe("userId");

    expect(wrapper.find(MaterialTable).at(0).props().columns[1].title).toBe("username");
    expect(wrapper.find(MaterialTable).at(0).props().columns[1].field).toBe("userName");

    expect(wrapper.find(MaterialTable).at(0).props().columns[2].title).toBe("role_label");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].field).toBe("role");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].hidden).toBeTruthy();

    expect(wrapper.find(EditProfilePopup).at(0).props().show).toBeFalsy();
    expect(wrapper.find(EditProfilePopup).at(0).props().user).toStrictEqual({
        userId: 0,
        userName: "",
      });
      act(() => {
        wrapper
          .find(EditProfilePopup)
          .at(0)
          .props()
          .action(true);
      });
      wrapper.render();
        wrapper.update();
    expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual([]);
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].title).toBe("userid");
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].field).toBe("userId");

    expect(wrapper.find(MaterialTable).at(0).props().columns[1].title).toBe("username");
    expect(wrapper.find(MaterialTable).at(0).props().columns[1].field).toBe("userName");

    expect(wrapper.find(MaterialTable).at(0).props().columns[2].title).toBe("role_label");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].field).toBe("role");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].hidden).toBeTruthy();

    expect(wrapper.find(EditProfilePopup).at(0).props().show).toBeTruthy();
    expect(wrapper.find(EditProfilePopup).at(0).props().user).toStrictEqual({
        userId: 0,
        userName: "",
      });
  });
  test("editUserProfile functionality", () => {
    let  wrapper = mount(<UsersTable/>) ;
    act(() => {
        wrapper.find(MaterialTable).props().actions[0].onClick({ target: { value: ''} } , {
            userId: 1,
            userName: "testedValue",
            roles:["Admin"]
          })
      });
      wrapper.update();
      wrapper.render();
    expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual([]);
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].title).toBe("userid");
    expect(wrapper.find(MaterialTable).at(0).props().columns[0].field).toBe("userId");

    expect(wrapper.find(MaterialTable).at(0).props().columns[1].title).toBe("username");
    expect(wrapper.find(MaterialTable).at(0).props().columns[1].field).toBe("userName");

    expect(wrapper.find(MaterialTable).at(0).props().columns[2].title).toBe("role_label");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].field).toBe("role");
    expect(wrapper.find(MaterialTable).at(0).props().columns[2].hidden).toBeTruthy();

    expect(wrapper.find(EditProfilePopup).at(0).props().show).toBeTruthy();
    expect(wrapper.find(EditProfilePopup).at(0).props().user).toStrictEqual({
      userId: 1,
      userName: "testedValue",
      roles:["Admin"]
    });
    expect(wrapper.find(EditRolePopup).at(0).props().show).toBeFalsy();
    expect(wrapper.find(EditRolePopup).at(0).props().user).toStrictEqual({
      userId: 1,
      userName: "testedValue",
      roles:["Admin"]
    });
  });

  test("composeTableActions functionality ", () => {
    const mockSuccessResponse = [{id:1 , name:"Name A"} , {id:2 , name:"Name B"}];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise);
    const mockJson = "SuperAdmin";
    const getItemMock =  jest.spyOn(Storage.prototype, "getItem").mockImplementation(() => mockJson);
    let  wrapper = mount(<UsersTable/>) ;
    // expect(getItemMock).toHaveBeenCalledTimes(2);
    // expect(getItemMock.mock.calls[0][0]).toBe("userRole");
    expect(wrapper.find(MaterialTable).props().actions.length).toBe(2)
    act(() => {
      wrapper.find(MaterialTable).props().actions[1].onClick({ target: { value: ''} } , {
          userId: 1,
          userName: "testedValue",
          roles:["Admin"]
        })
    });
    wrapper.update();
    wrapper.render();
    expect(wrapper.find(EditRolePopup).at(0).props().show).toBeTruthy();
    expect(wrapper.find(EditRolePopup).at(0).props().user).toStrictEqual({
      userId: 1,
      userName: "testedValue",
      roles:["Admin"]
    });
    let flag1=false;
    let show1 = false;
    act(() => {
      wrapper.find(EditRolePopup).props().action(show1,flag1)
    });
    wrapper.update();
    wrapper.render();
    expect(wrapper.find(EditRolePopup).at(0).props().show).toBeFalsy();
    return mockJsonPromise.then(() => {
      let flag2=true;
    let show2 = true;
      act(() => {
        wrapper.find(EditRolePopup).props().action(show2,flag2)
      });
    }).then(() => {
      wrapper.render();
      wrapper.update();
      expect(fetchMock).toHaveBeenCalledTimes(3);
      expect(fetchMock.mock.calls[0][0]).toBe("/api/user/lessRoleUsers");
      expect(fetchMock.mock.calls[1][0]).toBe("/api/user/1");
      expect(fetchMock.mock.calls[2][0]).toBe("/api/user/lessRoleUsers");
      expect(wrapper.find(EditRolePopup).at(0).props().show).toBeTruthy();
      expect(wrapper.find(MaterialTable).at(0).props().data).toStrictEqual(mockSuccessResponse);
    });
  });
  });
