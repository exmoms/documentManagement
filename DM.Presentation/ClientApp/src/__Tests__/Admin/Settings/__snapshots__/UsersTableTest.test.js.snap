// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable matches snapshot 1`] = `
<UsersTable
  i18n={Object {}}
  t={[Function]}
  tReady={false}
/>
`;
